### PR TITLE
feat(core): opt-in episodic-context recall section (episode-grouped raw source turns)

### LINKED ISSUE #2331: [harmonic 3/5] Episode-grouped recall: inject raw source segments for top recalled facts
Part of the **harmonic-memory** series (labels: `harmonic-memory` + `priority:P1`..`priority:P5`). This issue is **P3 — implement THIRD**. It is independent of P1 and P2.

### Series

| Order | Issue | Depends on |
|---|---|---|
| P1 | #2329 Construction pipeline | — |
| P2 | #2330 Merge-on-write | — (independent, but shares `extraction-persist.ts` touchpoints with #2329 — rebase carefully if both are in flight) |
| P3 | #2331 Episode-grouped recall | — |
| P4 | #2332 Deep-recall surface | **#2329 must merge first** (needs a populated anchor graph) |
| P5 | #2333 Span-mode extraction | — (bench-gated; do last) |

Order rationale: P1 makes the already-shipped harmonic retrieval tier functional and unblocks P4. P2 is the paper's largest measured quality gain. P3 is cheap and independent. P4 needs P1's data to be worth anything. P5 changes the extraction contract and only proceeds on benchmark evidence.

## Background

The Memora paper ([arXiv 2602.03315](https://arxiv.org/abs/2602.03315)) ablation (Table 4) shows that injecting the **raw source conversation segment** alongside extracted facts beats facts-only recall: LoCoMo LLM-judge 0.863 for "Episodic (Segment) + Factual" vs 0.833 for "Factual only", and raw segments beat *extracted summaries* of the same segments (0.838). The narrative "connective tissue" around a fact grounds multi-hop and temporal reasoning that isolated facts lose.

Remnic already stores everything needed and never re-surfaces it grouped:

- **Raw turns**: the LCM archive (`packages/remnic-core/src/lcm/archive.ts`) keeps every turn in SQLite (`lcm_messages`: `session_id`, `turn_index`, `role`, `content`, ...). Range fetch exists: `getMessages(sessionId, fromTurn, toTurn)` at `archive.ts:169-181`. FTS with content: `searchWithContent(query, limit, sessionId?, sessionPrefix?)` at `archive.ts:282-362` (returns message ids + excerpts).
- **Fact → source links**: persisted facts carry a `sources` frontmatter array of `ProvenanceSource` records — `sessionKey`, `turnId`, `observedAt`, `quote`, `charStart`, `charEnd` (`provenance.ts:50-82`, `types.ts:2790-2853`); extraction populates them at `extraction-persist.ts:232-290`.
- **Recall pipeline**: sections are registered in `buildDefaultRecallPipeline` (`config.ts:4258-4550`), gated via `isRecallSectionEnabled` / `getRecallSectionNumber` (`recall-section-coordinator.ts:78-126`), appended via `appendRecallSection`, and assembled + bounded at `recall-internal.ts:4921-4938` (`recallBudgetChars` = `maxMemoryTokens * 4` by default, `config.ts:4510-4523`).

## Goal

A new default-off recall section, `episodic-context`: after the core memory results are known, take the top recalled facts, resolve their `sources` provenance to LCM turn ranges, group overlapping ranges into episodes, fetch the raw turns, and inject up to N compact "Source Episodes" blocks into the recall context — within the existing recall budget and deadlines.

## Non-goals

- No handling of the legacy scalar `source: string` frontmatter field. Facts without a structured `sources` array are skipped (state this in docs). No string parsing heuristics.
- No LLM calls. This section is pure lookup + formatting; it must be cheap enough for the hot path.
- No new storage. Read-only against the LCM archive.
- No cross-plugin/archive backends.

## Config

1. Boolean flag `episodicContextEnabled` (default absent/false), parsed strict `cfg.episodicContextEnabled === true` next to `harmonicRetrievalEnabled` (`config.ts:2484-2486`); add to `PluginConfig` in `types.ts` (near line 1393) and to `openclaw.plugin.json` configSchema (both copies) + `docs/config-reference.md`.
2. Section registry entry in `buildDefaultRecallPipeline` (the array at `config.ts:4258-4550`), reusing **existing** `RecallSectionConfig` numeric keys so `getRecallSectionNumber` works unchanged:

```ts
{
  id: "episodic-context",
  enabled: cfg.episodicContextEnabled === true,
  maxResults: 2,   // max episodes injected
  maxTurns: 8,     // max raw turns per episode (paper: episodes run 2-8 messages)
  maxChars: 2400,
},
```

Zero semantics (Checklist #33): `maxResults: 0` or `maxTurns: 0` disables the section (short-circuit before any archive query); the schema must allow 0.

## Architecture

```mermaid
flowchart LR
  A[core memories resolved] --> B[collect ProvenanceSource from top-k recalled facts]
  B --> C["group by sessionKey; merge overlapping turn windows [start, end)"]
  C --> D["LcmArchive.getMessages(sessionId, fromTurn, toTurn)"]
  D --> E[clean + trim turns]
  E --> F["formatEpisodicContext() -> '## Source Episodes'"]
  F --> G[appendRecallSection episodic-context]
```

### Step 1 — Placement in `orchestration/recall-internal.ts`

This section **depends on the core memory results**, unlike the enrichment sections (e.g. harmonic at lines 1507–1590) which run in parallel off the query alone. Place it in the final phase, after the primary memory results are final and **before** `assembleRecallSections` (~line 4921). Wrap the whole computation in `runRecallSectionWithinDeadline` (`recall-qos.ts:180`) with `fallback: null` and a deadline derived from the remaining enrichment budget (`recallEnrichmentDeadlineMs`, default 25000 — `config.ts:3113-3120`; resolve remaining time the same way `resolveRecallCoreSectionDeadlineMs` at `recall-qos.ts:139` does). Emit a `recordRecallSectionMetric` entry (`recall-qos.ts:28-36`) with `section: "episodicContext"` on every exit path, including the disabled path — copy the harmonic section's metric discipline exactly.

Gate: `this.deps.isRecallSectionEnabled("episodic-context", false)` — note the **`false` default**; this section is opt-in, unlike most sections that default enabled (registry parser defaults `entry.enabled` to true, `config.ts:4224-4256`, which is why the registry entry above must carry the explicit flag expression).

### Step 2 — Provenance → episode windows (pure function, new module)

New file `packages/remnic-core/src/episodic-context.ts` (host-agnostic core):

```ts
export interface EpisodeWindow {
  sessionKey: string;
  fromTurn: number;      // inclusive
  toTurn: number;        // EXCLUSIVE — [start, end) semantics throughout (Checklist #23)
  factRank: number;      // best (lowest) rank among contributing facts — drives ordering
  memoryIds: string[];   // sorted, for the citation line
}

export function planEpisodeWindows(options: {
  recalledFacts: Array<{ memoryId: string; rank: number; sources?: ProvenanceSource[] }>;
  maxEpisodes: number;
  maxTurnsPerEpisode: number;
}): EpisodeWindow[];
```

Rules:

- Resolve each `ProvenanceSource.turnId` to a numeric turn index. **First verify the `turnId` format** against what extraction writes (`extraction-persist.ts:232-290`) — if it encodes the LCM `turn_index`, parse it directly. If a source's turnId cannot be resolved to a number, skip that source (do not guess).
- Window per source: `[turnIndex - 1, turnIndex + 2)` clamped to ≥ 0 — one turn of leading and trailing context.
- Merge overlapping/adjacent windows within the same `sessionKey`; cap each merged window's span at `maxTurnsPerEpisode` (trim the tail, keep the window containing the highest-ranked fact's turn).
- Order: `factRank asc`, then `sessionKey asc`, then `fromTurn asc` — a total, stable comparator that returns 0 for equal keys (Checklist #12). Take the first `maxEpisodes`.
- Deterministic output for shuffled input (test this).

### Step 3 — Fetch + clean

For each window, `LcmArchive.getMessages(sessionKey, fromTurn, toTurn - 1)` (**check `getMessages` bound semantics at `archive.ts:169-181` first** — if its `toTurn` is inclusive, pass `toTurn - 1`; keep `[start, end)` in our own types regardless).

Fallback when turn resolution failed for ALL sources of a fact but a `quote` exists: `searchWithContent(quote, 1, sessionKey)` (`archive.ts:282-362`) to locate the message, then window around its `turn_index`. This is the only fallback; if it misses, the fact contributes no episode.

Cleaning (raw turns contain injected context):

- Only `role` `user`/`assistant` rows; drop tool/system rows and all `lcm_message_parts` payloads.
- Run user rows through the existing `cleanUserMessage()` used by the turn-buffer path (locate it where `processTurn` cleaning happens; do not write a second cleaner — reuse).
- Truncate each turn to 300 chars with a `…` marker; truncate the whole section to the registry `maxChars`.

### Step 4 — Formatter (rule: never fork formatting)

Add `formatEpisodicContext(episodes)` to `orchestration/recall-result-formatter.ts`, following `formatHarmonicRetrievalResults` (lines 332–363):

```
## Source Episodes

### Episode: <sessionKey shortened to 8 chars>… turns <fromTurn>-<toTurn-1> (supports [<memoryId>, …])
user: <cleaned text>
assistant: <cleaned text>
```

Return `null` when there are zero episodes so the section is omitted entirely (no empty headers).

### Namespace / privacy rule (Checklist #30)

Only fetch sessions the caller may see. The provenance `sessionKey` values on a fact were written by the same principal's extraction inside the same storage namespace, but defense-in-depth is required: apply the same session-scoping the LCM search tool applies (see the `sessionPrefix` parameter on `LcmArchive.search`/`searchWithContent` and how the registered `engram_context_search` tool scopes it, `lcm/tools.ts:57-86`). If the current recall's namespace resolution provides a session prefix, pass it; a `sessionKey` outside the caller's scope is skipped. Add the two-namespace leakage test.

## Tests

New `packages/remnic-core/src/episodic-context.test.ts` + a recall-internal integration case:

1. `planEpisodeWindows`: overlap merge, `[start, end)` boundary cases — two facts whose windows share exactly one boundary turn must not double-count it (Checklist #23); span cap trimming; ordering determinism under input shuffle; `maxEpisodes`/`maxTurnsPerEpisode` = 0 → `[]` **before** any lookup.
2. Turn resolution: unparseable `turnId` → source skipped; quote fallback path hit when all turnIds fail; quote fallback miss → no episode (assert on concrete non-empty expectations for the positive cases — Checklist #6).
3. Formatter: zero episodes → `null`; truncation applied; only user/assistant roles rendered.
4. Integration: flag off → section absent and **zero LCM queries** (assert archive mock not called); flag on with a seeded archive → section present under `## Source Episodes` and total recall context still within `recallBudgetChars`.
5. Cross-namespace: fact with a foreign-session provenance under scoping → episode skipped.

## Verification gates

- `npm run preflight:quick`
- `npm run test:entity-hardening` (touches `config.ts` and `orchestration/`)
- Bench evidence: `@remnic/bench` **LongMemEval** (multi-session and temporal-reasoning categories are where the paper shows the biggest segment-context gains) with `episodicContextEnabled` on vs off; post the table.

## Acceptance criteria

- [ ] Flag off: recall byte-identical to today; no archive reads.
- [ ] Flag on: top recalled facts with structured provenance produce grouped raw-turn episodes in the context, deduped, ordered, budget-bounded, within the enrichment deadline.
- [ ] Facts without structured `sources` are silently skipped (documented).
- [ ] All tests + gates green; bench table posted.

## Notes for the implementing agent

- Public repo: synthetic conversations only in fixtures; persona paths (`/home/user/...`) if a path is ever needed.
- Read the harmonic enrichment section (`recall-internal.ts:1507-1590`) once before starting — it is the template for metrics, gating, and abort discipline; this issue's section differs only in *when* it runs (after core results) and in having no LLM/abort handle.
- Keep the PR to this one section. Do not refactor the section coordinator.

### PR DESCRIPTION
**Pull request by joshuaswarren**
Fixes #2331

## What

Default-off `episodicContextEnabled` adds a `## Source Episodes` recall section. After the primary memory results are final (and before assembly), the section resolves the structured `sources` provenance of the top recalled facts to LCM archive turn ranges, groups overlapping ranges per session into episodes, and injects up to 2 episodes of cleaned raw user/assistant turns (max 8 turns each, 2400-char section cap) inside the existing recall budget and enrichment deadline.

Per the Memora ablation (issue table): raw source segments injected alongside facts beat facts-only recall; segments beat extracted summaries.

## Implementation

- `packages/remnic-core/src/episodic-context.ts` — pure planning: `planEpisodeWindows` (`[start, end)` semantics, overlap+adjacency merge, span cap anchored on the best-ranked fact's turn, total stable ordering, shuffle-deterministic), `resolveProvenanceTurnIndex` (parses the trailing turn index of the host turn fingerprint; unparseable → skip, never guess), `planEpisodeWindowsWithFallback` (single documented fallback: quote lookup via injected locator when ALL of a fact's turnIds fail).
- `packages/remnic-core/src/orchestration/episodic-context-section.ts` — section assembly (sibling module; the watched `recall-internal.ts` ceiling is unchanged): reads top-8 recalled facts' provenance, plans, fetches via `LcmEngine.getTurnRange` (new read-only inclusive fetch), filters user/assistant rows, cleans user rows through the shared cleaner, 300-char per-turn cap, appends within `maxChars`. Wrapped in `runRecallSectionWithinDeadline` (remaining enrichment budget resolved like `resolveRecallCoreSectionDeadlineMs`); `episodicContext` metric on every exit path incl. `skip(limit=0)` / `skip(no-memories)` / `timeout`.
- `LcmEngine.getTurnRange` — thin read-only passthrough to `LcmArchive.getMessages`; no token reshaping.
- Cleaner reuse, not a fork: the generic `cleanUserMessageWithPattern` moved to `@remnic/core/user-message-cleaning`; the OpenClaw wiring (`src/user-message-cleaning.ts`) keeps its channel-envelope pattern and delegates.
- Formatter: `RecallResultFormatter.formatEpisodicContext` returns `null` for zero episodes (no empty headers); `### Episode: <8-char key>… turns a-b (supports [ids])`.
- Namespace defense-in-depth (Checklist #30): candidate LCM session ids are framed ONLY under the caller's authorized `lcmReadNamespaces`; a foreign session's rows live under a foreign sentinel frame and are unreachable. Covered by a cross-namespace test.
- Config: strict `cfg.episodicContextEnabled === true` parse next to `harmonicRetrievalEnabled`; registry entry `(maxResults: 2, maxTurns: 8, maxChars: 2400)` after `memories`; `0` for `maxResults`/`maxTurns` disables before any archive query. `PluginConfig` type + all THREE `configSchema` copies (root, plugin-openclaw, shim-openclaw-engram) + `docs/config-reference.md`.

## Non-goals honored

No legacy scalar `source` parsing (facts without structured `sources` are silently skipped — documented). No LLM calls. No new storage. Read-only archive access. No cross-backend reach.

## Tests

`episodic-context.test.ts` (11): fingerprint/turnId resolution incl. rejection cases; zero-limit short-circuit → `[]` before any lookup; skip of sourceless facts and unparseable turnIds; overlap + boundary-adjacency merge (no turn double-counting); span cap keeps the best-ranked fact's turn inside incl. `maxTurns=1`; shuffle determinism + ordering contract; `maxEpisodes` by rank; quote-fallback hit/miss/throw; formatter null-on-empty + rendering; cleaner contract (leading preamble + wrapper strip, mid-message title preserved).

`episodic-context-recall.integration.test.ts` (3, real Orchestrator + real SQLite LCM archive): flag off → no section AND zero archive reads (counted via engine read wrappers); flag on → merged episode injected with raw turns, sourceless fact absent, within budget envelope; foreign-session provenance skipped.

## Gates (receipts in CI)

- `npm run test:entity-hardening` — 340/340 pass (touches `config.ts` + `orchestration/`)
- `npm run preflight:quick` — OK (quick), exit 0
- `node scripts/check-ratchets.mjs` — OK (`recall-internal.ts` net +0: section extracted to a sibling module; `orchestrator.ts` net +0; `types.ts` net +0)
- `npm run check-config-contract` — OK

## Bench evidence

The issue's LongMemEval on/off table requires the full dataset plus an answering/judging LLM provider; neither is available in this environment (no `OPENAI_API_KEY`; dataset not vendored — the bundled fixture is a 1-item smoke). The recall-side behavior is covered by the integration tests above; the bench table should be produced by the operator harness that has dataset + provider access. Flagged honestly rather than faked.

---

> [!NOTE]
> **Medium Risk**
> Changes live recall assembly and reads LCM archives with namespace-scoped session keys; failures are fail-open but misconfiguration could add latency or leak wrong-session context if mapping were wrong (integration tests cover cross-namespace skip).
> 
> **Overview**
> Adds **default-off** `episodicContextEnabled` and a new `episodic-context` recall pipeline section that injects **`## Source Episodes`**: structured `sources` on top recalled facts are mapped to LCM turn windows, merged per session, and rendered as cleaned raw user/assistant turns within existing recall budget and enrichment deadlines (no LLM, no new storage).
> 
> **Recall path:** `appendEpisodicContextSection` runs after memory recall and before final assembly; when disabled it does not touch the archive. **Planning** lives in `episodic-context.ts` (overlap/adjacency merge, caps, quote fallback). **Fetch** uses new read-only `LcmEngine.getTurnRange`. **Formatting** via `formatEpisodicContext`; user turns go through shared `@remnic/core/user-message-cleaning` (OpenClaw host delegates instead of duplicating logic).
> 
> Config/schema/docs updated (`recallPipeline` defaults: `maxResults` 2, `maxTurns` 8, `maxChars` 2400). Unit and orchestrator integration tests cover flag gating, namespace isolation, and budget behavior.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 732add37439f4f47f2a2b565647715954653f8f9. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

## Summary by CodeRabbit

* **New Features**
  * Added optional episodic context to recall results, disabled by default.
  * Recall can include up to two grouped source episodes with relevant archived turns.
  * Added configuration support for enabling episodic context.
* **Improvements**
  * Archived user messages are cleaned for clearer source context.
  * Context respects existing turn, character, and result limits.
* **Documentation**
  * Updated configuration reference with episodic-context behavior and limits.
* **Tests**
  * Added coverage for provenance handling, episode grouping, limits, cleaning, and fallback behavior.

### CODE REVIEW

File: packages/remnic-core/src/orchestration/recall-internal.ts [hunk 1]
```diff
@@ -4926,6 +4914,13 @@ export class RecallInternalCoordinator {
       recallPlan: timings.recallPlan,
       queryPolicy: timings.queryPolicy,
     });
+    // Episodic context (#2331): opt-in raw-turn episodes for the top recalled
+    // facts; flag off is a no-op so recall output stays byte-identical.
+    await appendEpisodicContextSection(this.deps, {
+      sectionBuckets, recalledMemoryPaths, recalledMemoryNamespaces, scopePlan, namespacesEnabled,
+      enrichmentSectionDeadlineMs, recallOuterTimeoutMs, recallStart,
+      recordMetric: recordRecallSectionMetric, abortSignal: options.abortSignal,
+    });
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply packages/remnic-core/src/orchestration/recall-internal.ts]: Fixed in d2c5eb1a: recall-internal now derives index-aligned namespaces from xrayRecalledResults (populated by the same branches that populate paths) before the episodic call.

File: packages/remnic-core/src/orchestration/episodic-context-section.ts [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+          const fetched = await engine.getTurnRange(
+            sessionId,
+            window.fromTurn,
+            window.toTurn - 1,
+          );
+          if (fetched.length > 0) {
+            rows = fetched;
+            break;
+          }
+        }
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply packages/remnic-core/src/orchestration/episodic-context-section.ts:204]: Fixed in d2c5eb1a: per-candidate try/catch for getTurnRange and quote search inside the section, plus a fail-open wrapper at the recall-internal call site — an archive error degrades the section, never the recall.

File: packages/shim-openclaw-engram/openclaw.plugin.json [hunk 3]
```diff
@@ -105,7 +105,11 @@
     "type": "object",
     "additionalProperties": false,
     "properties": {
-      "abstractionAnchorsEnabled": {
+      "episodicContextEnabled": {
+      "label": "Episodic Context",
+      "help": "Inject episode-grouped raw source turns for the top recalled facts as a compact Source Episodes section."
+    },
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply packages/shim-openclaw-engram/openclaw.plugin.json]: Fixed in d2c5eb1a: malformed duplicate removed from all three manifest copies; the single remaining episodicContextEnabled entry is the valid boolean schema.

File: packages/remnic-core/src/user-message-cleaning.ts [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+  content: string,
+  platformHeaderPattern: RegExp,
+): string {
+  let cleaned = content;
+  // Remove structured host-injected memory wrappers wherever the platform
+  // emits them; free-form markdown stripping below is intentionally anchored.
+  cleaned = cleaned.replace(
+    /<supermemory-context[^>]*>[\s\S]*?<\/supermemory-context>\s*/gi,
+    "",
+  );
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply packages/remnic-core/src/user-message-cleaning.ts]: Fixed in d2c5eb1a: opening tag now requires whitespace-led attributes or immediate close — verified linear (1ms on 20k-repetition evil input); plain and attributed wrappers still strip.

File: packages/remnic-core/src/episodic-context.ts [hunk 5]
```diff
@@ -0,0 +1,272 @@
...
+  left: EpisodeWindow,
+  right: EpisodeWindow,
+): number {
+  return (
+    left.factRank - right.factRank ||
+    (left.sessionKey < right.sessionKey ? -1 : left.sessionKey > right.sessionKey ? 1 : 0) ||
+    left.fromTurn - right.fromTurn ||
+    left.toTurn - right.toTurn ||
+    (left.memoryIds < right.memoryIds ? -1 : left.memoryIds > right.memoryIds ? 1 : 0)
+  );
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply packages/remnic-core/src/episodic-context.ts:170]: Fixed in d2c5eb1a: compareEpisodeWindows joins memoryIds with a NUL separator before the relational compare.

File: packages/remnic-core/src/orchestration/recall-internal.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply packages/remnic-core/src/orchestration/recall-internal.ts]: Fixed in d2c5eb1a (same as the Cursor finding): namespaces are derived from xrayRecalledResults at the call site, index-aligned with recalledMemoryPaths.

File: packages/remnic-core/src/orchestration/recall-result-formatter.ts [hunk 6]
```diff
@@ -384,6 +384,37 @@ export class RecallResultFormatter {
...
+        episode.sessionKey.length > 8
+          ? `${episode.sessionKey.slice(0, 8)}…`
+          : episode.sessionKey;
+      const header =
+        `### Episode: ${shortKey} turns ${episode.fromTurn}-${episode.toTurn - 1} ` +
+        `(supports [${episode.memoryIds.join(", ")}])`;
+      const lines = episode.turns.map(
+        (turn) => `${turn.role}: ${turn.content}`,
+      );
+      return [header, ...lines].join("\n");
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply packages/remnic-core/src/orchestration/recall-result-formatter.ts:419]: Fixed in d2c5eb1a: formatEpisodicContext fences turn content via renderAuthorityBoundContent with unknown origin when originAuthority is enabled.

File: packages/shim-openclaw-engram/openclaw.plugin.json (hunk 3 above)
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply packages/shim-openclaw-engram/openclaw.plugin.json]: Fixed in d2c5eb1a: duplicate removed from all three manifests; the valid typed entry is kept.

File: packages/remnic-core/src/user-message-cleaning.ts [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+  let cleaned = content;
+  // Remove structured host-injected memory wrappers wherever the platform
+  // emits them; free-form markdown stripping below is intentionally anchored.
+  // The opening tag admits only whitespace-led attributes (or an immediate
+  // close) so repeated `<supermemory-context` literals cannot drive
+  // polynomial backtracking (CodeQL redos rule).
+  cleaned = cleaned.replace(
+    /<supermemory-context(?:\s[^>]*)?>[\s\S]*?<\/supermemory-context>\s*/gi,
+    "",
+  );
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply packages/remnic-core/src/user-message-cleaning.ts]: Fixed in fe1b1d492: the attribute run is bounded ({0,256}) and excludes angle brackets, capping per-position work; verified linear (1-2ms) on 20k-repetition inputs with and without attribute-space separators, with strip behavior unchanged.

File: packages/remnic-core/src/user-message-cleaning.ts [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+  // emits them; free-form markdown stripping below is intentionally anchored.
+  // The opening tag admits one whitespace-led attribute run, bounded (256
+  // chars) and exclusive of angle brackets, so repeated `<supermemory-context`
+  // literals cannot drive polynomial backtracking (CodeQL redos rule) — the
+  // attribute scan stops at the next `<` or `>` and the bound caps the work
+  // per start position.
+  cleaned = cleaned.replace(
+    /<supermemory-context(?:\s[^<>]{0,256})?>[\s\S]*?<\/supermemory-context>\s*/gi,
+    "",
+  );
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply packages/remnic-core/src/user-message-cleaning.ts]: Fixed in cfa3d537d + this head: bounded quantifiers still rescan per start position, so the strip is now a single left-to-right string-search pass with no regex on uncontrolled input at all — linear on all adversarial inputs (68ms vs 4.2s on the 20k-repeat worst case), strip semantics verified unchanged.

File: packages/remnic-core/src/user-message-cleaning.ts [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+      cursor = open + WRAPPER_OPEN.length;
+      continue;
+    }
+    const close = content.indexOf(WRAPPER_CLOSE, headEnd);
+    if (close === -1) return out + content.slice(cursor);
+    let after = close + WRAPPER_CLOSE.length;
+    while (after < content.length && isWhitespaceCode(content.charCodeAt(after))) {
+      after += 1;
+    }
+    cursor = after;
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply packages/remnic-core/src/user-message-cleaning.ts:59]: Fixed in 732add374: the matched-wrapper branch now appends content.slice(cursor, open) before advancing, so preceding and between-wrapper text is preserved; regression test added.

File: packages/remnic-core/src/user-message-cleaning.ts [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+
+  // Remove markdown memory context only when it is a leading preamble. If a
+  // user writes a section with this title later in their message, preserve it.
+  cleaned = cleaned.replace(
+    /^\s*## Memory Context \((?:Engram|Remnic)\)[\s\S]*?(?=\n## |\n$)/i,
+    "",
+  );
+
+  if (hasPlatformHeader) {
+    cleaned = cleaned.replace(/\s*\[message_id:\s*[^\]]+\]\s*$/i, "");
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply packages/remnic-core/src/user-message-cleaning.ts]: Fixed in 732add374: the trailing [message_id: …] strip is now a right-anchored string scan (stripTrailingMessageId) with no regex on uncontrolled input; adapter-scenario suite passes.
